packages/require/lib/index.ts

```diff
@@ -0,0 +1,128 @@
+
+export function file(options: RequireOptions) {
+  const sourceFilePath = path.resolve(options.file);
+
+  expectOptions(options, ["file"]);
+
+  const conf = TruffleConfig.default().with(options);
+
+  const scriptModule = new Module(sourceFilePath);
+
+  // Provides a subset of the globals listed here: https://nodejs.org/api/globals.html
+  const sandbox: { [index: string]: any } = {
+    __filename: path.basename(sourceFilePath),
+    __dirname: path.dirname(sourceFilePath),
+    Buffer,
+    clearImmediate,
+    clearInterval,
+    clearTimeout,
+    console,
+    exports: scriptModule.exports,
+    global,
+    process,
+    setImmediate,
+    setInterval,
+    setTimeout,
+    module: scriptModule,
+    config: conf,
+    artifacts: conf.resolver,
+    require: (id: string) => {
+      // Ugh. Simulate a full require function for the file.
+      id = id.trim();
+
+      // If absolute, just require.
+      if (path.isAbsolute(id)) return originalRequire(id);
+
+      // If relative, it's relative to the file.
+      if (id[0] === ".") {
+        return originalRequire(path.join(path.dirname(sourceFilePath), id));
+      } else {
+        // Not absolute, not relative, must be a globally or locally installed module.
+        // Try local first.
+        // Here we have to require from the node_modules directory directly.
+
+        var moduleDir = conf.working_directory;
+        while (true) {
+          try {
+            return originalRequire(path.join(moduleDir, "node_modules", id));
+          } catch (e) {}
+          var oldModuleDir = moduleDir;
+          moduleDir = path.dirname(moduleDir);
+          if (moduleDir === oldModuleDir) break;
+        }
+
+        // Try global, and let the error throw.
+        return originalRequire(id);
+      }
+    }
+  };
+
+  // Now add contract names.
+  Object.keys(conf.context || {}).forEach(key => {
+    sandbox[key] = conf.context[key];
+  });
+
+  const context = vm.createContext(sandbox);
+
+  const old_cwd = process.cwd();
+  const cwd = path.dirname(sourceFilePath);
+  process.chdir(cwd);
+
+  const source = compile(conf, sourceFilePath, context);
+
+  const script = new vm.Script(source, { filename: sourceFilePath });
+
+  script.runInContext(context);
+  scriptModule.loaded = true;
+
+  const returnValue = scriptModule.exports.default ?? scriptModule.exports;
+
+  process.chdir(old_cwd);
+
+  return returnValue;
+}
+
+export function exec(options: ExecOptions, done: (...args: any[]) => void) {
+  expectOptions(options, [
+    "contracts_build_directory",
+    "file",
+    "resolver",
+    "provider",
+    "network",
+    "networks",
+    "network_id"
+  ]);
+
+  const interfaceAdapter = createInterfaceAdapter({
+    provider: options.provider,
+    networkType: options.networks[options.network].type
+  });
+  const web3 = new Web3Shim({
+    provider: options.provider,
+    networkType: options.networks[options.network].type
+  });
+
+  try {
+    const fn = this.file({
```

Comment:
What is `this` here? The global scope?

Reply:
Good catch! I really don't know why it's like that. The only reason why this function is showing up as changed is because I decided to take this as an opportunity to port this entire module over to TS, and the only changes that I made to it were to support that change.

I have a sneaking suspicion that it's a very ugly ugly hack that someone put into place so that they could bind `this` to cause `exec` to call some local implementation of `file`, but I'm not certain. I'll dig into it and remove the `this` if I can.

Reply:
Weird. I'm fine with it as is. I was just wondering if you had already dug into it.

I'll let you mark this one as Resolved in case you want to look into it (you might want to `git mv` the old implementation to this new location, just in case there is some good history there).

Reply:
FWIW, I did do a `git mv` when I set this up, but it seems that enough of the file has changed that it didn't track its history. I'll see if I can insert a commit that does the move as a separate commit.